### PR TITLE
Separate context propagation into separate layers (otep-66)

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
This is a draft PR to incorporate the changes in [otep-66](https://github.com/open-telemetry/oteps/pull/66), specifically breaking context propagation into separate layers.

### CODE REVIEW

File: src/OpenTelemetry.Api/Baggage/BaggageContext.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Api.Baggage
+{
+    public abstract class BaggageContext
+    { }
```
Comment (lmolkova) [inline src/OpenTelemetry.Api/Baggage/BaggageContext.cs]: 1. why this abstraction is needed? Is there an existing .NET primitive that could be used instead? IEnumerable or Dictionary?
2. Does it have to be a class?
3. Why it does not have GEt/Set/Remove/Clear and we need a manager for that? This does not seem like a .NET friendly API

File: src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+using OpenTelemetry.Api.Context.Propagation;
+
+namespace OpenTelemetry.Api.Baggage
+{
+    public abstract class BaggageManagerBase
+    {
+        public abstract string GetEntry(BaggageContext context, string key);
+
+        public abstract BaggageContext SetEntry(BaggageContext context, string key, string value);
```
Comment (lmolkova) [inline src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs]: is it going to create a new BaggageContext or update one passed in the argument? If former, it seems inefficient. If latter, this should be changes to the method on BaggageContext

File: src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using OpenTelemetry.Api.Context.Propagation;
+
+namespace OpenTelemetry.Api.Baggage
+{
+    public abstract class BaggageManagerBase
```
Comment (lmolkova) [inline src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs]: why do we need this class?

File: src/OpenTelemetry.Api/Correlations/CorrelationContext.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Api.Correlations
+{
+    public class CorrelationContext
+    { }
```
Comment (lmolkova) [inline src/OpenTelemetry.Api/Correlations/CorrelationContext.cs]: same concerns as for the BaggageContext

File: src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        public abstract string GetEntry(BaggageContext context, string key);
+
+        public abstract BaggageContext SetEntry(BaggageContext context, string key, string value);
+
+        public abstract BaggageContext RemoveEntry(BaggageContext context, string key);
+
+        public abstract BaggageContext ClearEntries(BaggageContext context);
+
+        public abstract HttpInjectorBase GetHttpInjector();
```
Comment (lmolkova) [inline src/OpenTelemetry.Api/Baggage/BaggageManagerBase.cs]: Is there a requirement to propagate extractors/injectors through the manager class? Could it be done as a constructor argument to adapters instead?

How extractor/injections are going to be set up by users? On the tracer/meter factory?

**Comment by MikeGoldsmith** [comment]
Thanks for the initial review @lmolkova. This is not complete yet and I just wanted to push something so show it's in progress - that's why I set the PR as _draft_.

The work I've done so far in the above commits are trying to stitch together what's been done in Go & Java and will be refactored to better suit OTel .NET in the coming days.

**Pull request closed without merging**